crates/driver/src/util/math.rs

```diff
@@ -0,0 +1,18 @@
+use ethereum_types::U256;
+
+/// Computes `x * q / d` rounding down.
+///
+/// Returns `None` if `d` is `0` or if the result overflows a 256-bit integer.
+pub fn mul_ratio(x: U256, q: U256, d: U256) -> Option<U256> {
+    x.full_mul(q).checked_div(d.into())?.try_into().ok()
+}
+
+/// Computes `x * q / d` rounding up.
+///
+/// Returns `None` if `d` is `0` or if the result overflows a 256-bit integer.
+pub fn mul_ratio_ceil(x: U256, q: U256, d: U256) -> Option<U256> {
+    let p = x.full_mul(q);
+    let result = U256::try_from(p.checked_div(d.into())?).ok()?;
+    let round_up = !p.checked_rem(d.into())?.is_zero();
+    result.checked_add((round_up as u8).into())
+}
```

Comment:
Would those warrant a unit test?

Reply:
I can add one if you like. The code was simple enough that I didn’t think it warranted one (I put it in a math utility function more for code legibility than anything else). Once I finish implementing the E2E test, this code should get coverage (and technically already does in the unit tests added in #1861).